File: checker.py
import requests
import random
import time
from colorama import Fore

aqua = """
                                                                                                      
██████╗ ██╗   ██╗██████╗ ██████╗  ██████╗ 
██╔══██╗██║   ██║██╔══██╗██╔══██╗██╔═══██╗
██████╔╝██║   ██║██████╔╝██████╔╝██║   ██║
██╔══██╗██║   ██║██╔══██╗██╔══██╗██║   ██║
██████╔╝╚██████╔╝██████╔╝██████╔╝╚██████╔╝
╚═════╝  ╚═════╝ ╚═════╝ ╚═════╝  ╚═════╝                                           
--------------------------------------------------                                                                                                                                                                                                                                                                                
    """

print(Fore.BLUE + f"{aqua}" + Fore.RESET)

def is_taken(user):
    try:
        with open("Status/Taken.txt", "r") as file:
            taken_usernames = file.read().splitlines()
    except FileNotFoundError:
        taken_usernames = []
    return user in taken_usernames

def add_to_taken(user):
    with open("Status/Taken.txt", "a") as file:
        file.write(user + "\n")

def generate_username():
    characters = "abcdefghijklmnopqrstuvwxyz123456789"
    while True:
        include_hyphen = random.choice([True, False])
        
        if include_hyphen:
            part = "".join(random.choices(characters, k=2)) 
            hyphen_position = random.randint(1, len(part) - 1) 
            username = part[:hyphen_position] + '-' + part[hyphen_position:]
        else:
            username = "".join(random.choices(characters, k=3))
        
        if username[0] != '-' and username[-1] != '-' and '--' not in username:
            return username

while True:
    user = generate_username()
    if is_taken(user):
        continue
    response = requests.get(f"https://www.github.com/{user}/")

    if response.status_code == 200:
        print(Fore.RED + f"USERNAME TAKEN: {user}" + Fore.RESET)
        add_to_taken(user)
    elif response.status_code == 404:
        print(Fore.GREEN + f"USERNAME AVAILABLE: {user}" + Fore.RESET)
        with open("Status/NotTaken.txt", "a") as file:
            file.write(user + "\n")         
    else:
        print("BLOCKED FROM GITHUB")
        time.sleep(5)

File: checker.js
require('dotenv').config();
const axios = require('axios');
const fs = require('node:fs');
const path = require('node:path');
const { randomInt } = require('node:crypto');

const Taken = path.join(__dirname, 'Status', 'Taken.txt');
const NotTaken = path.join(__dirname, 'Status', 'NotTaken.txt');

const Hyphen = process.env.HYPHEN === 'true';
const Numbers = process.env.NUMBERS === 'true';

const base = 'abcdefghijklmnopqrstuvwxyz';
const numbers = '0123456789';
const characters = Numbers ? `${base}${numbers}` : base;

let length = Number.parseInt(process.env.LENGTH, 10) || 3; 

const loadTakenUsernames = () => {
    const takenUsernames = new Set();
    try {
        const data = fs.readFileSync(Taken, 'utf8');
        for (const user of data.split('\n').filter(Boolean)) {
            takenUsernames.add(user.trim());
        }
        console.log(`Loaded ${takenUsernames.size} taken usernames.`);
    } catch (err) {
        if (err.code !== 'ENOENT') throw err;
    }
    return takenUsernames;
};

const isUsernameTaken = (username, takenUsernames) => {
  return takenUsernames.has(username);
};

const generateUsername = () => {
    if (length < 2 || length > 5) {
        console.error(`Unsupported username length: ${length}. Using default length 3.`);
        length = 3; 
    }

    while (true) {
        let username;
        let baseLength = length - (Hyphen ? 1 : 0);

        if (baseLength < 1) {
            baseLength = 1;
        }

        username = [...Array(baseLength)].map(() => characters.charAt(randomInt(0, characters.length))).join('');

        if (Hyphen) {
            const hyphenPosition = randomInt(1, username.length);
            username = `${username.slice(0, hyphenPosition)}-${username.slice(hyphenPosition)}`;
        }

        if (username[0] !== '-' && username[username.length - 1] !== '-' && !username.includes('--')) {
            return username;
        }
    }
}; 

const checkUsernameAvailability = async (username) => {
    while (true) {
        try {
            const response = await axios.get(`https://www.github.com/${username}/`);
            if (response.status === 200 || response.status === 410) {
                return 'taken';
            }
            if (response.status === 404) {
                return 'available';
            }
        } catch (err) {
            if (err.response) {
                if (err.response.status === 429) {
                    console.log(`Rate limit exceeded for ${username}. Waiting 4 seconds before retrying...`);
                    await new Promise(resolve => setTimeout(resolve, 4000));
                } else if (err.response.status === 410) {
                    return 'taken';
                } else if (err.response.status === 404) {
                    return 'available';
                }
                console.error(`Error checking username ${username}: ${err.response.status}`);
                throw err;
            }
            console.error(`Error checking username ${username}: ${err.message}`);
            throw err;
        }
    }
};

const checkUsernames = async () => {
    const takenUsernames = loadTakenUsernames();

    while (true) {
        const username = generateUsername();

        if (isUsernameTaken(username, takenUsernames)) {
            continue;
        }

        try {
            const result = await checkUsernameAvailability(username);

            if (result === 'taken') {
                console.log(`\x1b[31mUSERNAME TAKEN: ${username}\x1b[0m`);
                fs.appendFileSync(Taken, `${username}\n`);
                takenUsernames.add(username); 
            } else if (result === 'available') {
                console.log(`\x1b[32mUSERNAME AVAILABLE: ${username}\x1b[0m`); 
                fs.appendFileSync(NotTaken, `${username}\n`);
            }
        } catch (err) {
            console.error(`Error processing username ${username}: ${err.message}`);
            await new Promise(resolve => setTimeout(resolve, 5000));
        }
    }
};

checkUsernames();
